### PR TITLE
State token in callback query is checked against state token in cookie

### PR DESCRIPTION
**Pull request by Sean-Q-Sun**
For Issue #53 
@ericchiang Please take a look

### CODE REVIEW

**Comment by coreos-ui** [comment]
Can one of the admins verify this patch?

**Comment by coreos-ui** [comment]
Can one of the admins verify this patch?

**Comment by ggreer** [comment]
jenkins rebuild

**Comment by ggreer** [comment]
jenkins rebuild

File: auth/auth.go [hunk 1]
```diff
@@ -25,14 +25,17 @@ import (
...
+	CSRFHeader        = "X-CSRFToken"
+	stateCookieName   = "state-token"
+	errorOAuth        = "oauth_error"
+	errorLoginState   = "login_state_error"
+	errorCookie       = "cookie_error"
+	errorInternal     = "internal_error"
+	errorMissingCode  = "missing_code"
+	errorMissingState = "missing_state"
+	errorInvalidCode  = "invalid_code"
+	errorInvalidState = "invalid_state"
```
Comment (ericchiang) [inline auth/auth.go]: These new error messages new associated frontend descriptions here: https://github.com/openshift/console/blob/417be1e1587415f9c364bb26090e744e24d14938/frontend/public/components/error.jsx

File: auth/auth.go [hunk 2]
```diff
@@ -262,6 +273,14 @@ func (a *Authenticator) CallbackFunc(fn func(loginInfo LoginJSON, successURL str
 		q := r.URL.Query()
 		qErr := q.Get("error")
 		code := q.Get("code")
+		urlState := q.Get("state")
+
+		cookieState, err := r.Cookie(stateCookieName)
+		if err != nil {
+			log.Infof("missing state in StateCookie: %v", err)
```
Comment (ericchiang) [inline auth/auth.go]: nit: instead of Infof use Errorf. Also maybe "failed to parse state cookie"

File: auth/auth.go [hunk 3]
```diff
@@ -275,6 +294,12 @@ func (a *Authenticator) CallbackFunc(fn func(loginInfo LoginJSON, successURL str
 			return
 		}
 
+		if urlState != cookieState.Value {
+			log.Errorf("State in url does not match State cookie")
+			a.redirectAuthError(w, errorInvalidState, nil)
+
```
Comment (ericchiang) [inline auth/auth.go]: nit: remove space

**Comment by ericchiang** [comment]
Also your commit email looks messed up

```
commit 643d2bac2f4dc2873febb114759fc96d831a441d
Author: Sean Sun <seansun@Seans-MBP.sfo.coreos.systems>
Date:   Tue May 22 14:37:15 2018 -0700

    State token in callback query is checked against state token in cookie
```

**Comment by ggreer** [comment]
jenkins rebuild

File: auth/auth.go (hunk 1 above)
Comment (ericchiang) [reply auth/auth.go]: @Sean-Q-Sun still working on this bit ^ ?

Comment (Sean-Q-Sun) [reply auth/auth.go]: Just added!

**Comment by coreos-ui** [comment]
Can one of the admins verify this patch?

**Comment by coreos-ui** [comment]
Can one of the admins verify this patch?

**Pull request merged**